src/network/netdev/netns.c

```diff
@@ -0,0 +1,130 @@
+#include "netlink-util.h"
+#include "networkd-manager.h"
+#include "path-util.h"
+
+#define NETNS_RUN_DIR "/var/run/systemd/netif/netns"
+
+static int setup_ns_path(NetDev *netdev, char **path) {
+        _cleanup_free_ char *ns_path = NULL;
+
+        ns_path = path_join(NETNS_RUN_DIR, netdev->netns);
+        if (!ns_path)
+                return log_oom();
+
+        *path = TAKE_PTR(ns_path);
+
+        return 0;
+}
+
+int setup_netns(NetDev *netdev, char *ns_path) {
+        _cleanup_close_ int netns = -1, fd = -1;
+        int r;
+
+        assert(netdev);
+
+        if (unshare(CLONE_NEWNET) < 0) {
+                r = -errno;
+                goto fail;
+        }
+
+        netns = open("/proc/self/ns/net", O_RDONLY|O_CLOEXEC|O_NOCTTY);
+        if (netns < 0) {
+                r = -errno;
+                goto fail;
+        }
+
+        fd = open(ns_path, O_RDONLY|O_CREAT|O_EXCL, 0);
+        if (fd < 0) {
+                if (errno != EEXIST) {
+                        r = -errno;
+                        goto fail;
+                } else
+                        return 0;
+        }
+
+        r = mount("/proc/self/ns/net", ns_path, "none", MS_BIND, NULL);
+        if (r < 0) {
+                log_error_errno(r, "Failed to bind /proc/self/ns/net: %m");
+                goto fail;
+        }
+
+ fail:
+        return r;
+}
+
+/* callback for moving netdev to namespace */
+static int move_netdev_namespace_handler(sd_netlink *rtnl, sd_netlink_message *m, NetDev *netdev) {
+        int r;
+
+        assert(netdev);
+        assert(netdev->state != _NETDEV_STATE_INVALID);
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0) {
+                log_netdev_warning_errno(netdev, r, "Netdev could not be moved to namespace : %m");
+                return 1;
+        }
+
+        log_netdev_debug(netdev, "Moved to namespace");
+
+        return 1;
+}
+
+int netdev_move_namespace(NetDev *netdev, char *ns_path) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL, *reply = NULL;
+        _cleanup_close_ int fd = -1;
+        int r;
+
+        r = sd_rtnl_message_new_link(netdev->manager->rtnl, &m, RTM_NEWLINK, netdev->ifindex);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to allocate generic netlink message: %m");
+
+        r = sd_rtnl_message_link_set_nlmsg_flags(m, NLM_F_REQUEST);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to set netlink message flag: %m");
+
+        fd = open(ns_path, O_RDONLY);
+        if (fd < 0)
+                return log_netdev_error_errno(netdev, -errno, "Failed open path : %m");
+
+        r = sd_netlink_message_append_u32(m, IFLA_NET_NS_FD, fd);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append IFLA_NET_NS_FD attribute: %m");
+
+        r = netlink_call_async(netdev->manager->rtnl, NULL, m, move_netdev_namespace_handler, NULL, netdev);
```

Comment:
destroy callback is not necessary since we not going to drop the netdev if it does not move into the namespace.

Reply:
It may not be necessary, but still I prefer to enable destroy callback.

If you really do not like that, then please drop `netdev_ref()` below, otherwise the `NetDev` object will not be freed.